### PR TITLE
[GPU] Fix command buffer support for cuDNN fusions.

### PR DESCRIPTION
**Pull request by sergachev**
CuDnnCmd is constructed before DnnGraph in CuDnnThunk is initialized so CuDnnCmd has to get  unique_ptr\<DnnGraph\>& instead of DnnGraph& at initialization.

Accordingly cuDNN thunks have to be initialized before command buffer ones to initialize graphs before they get captured.

Test CommandBuffersAreSupported used to not demonstrate the use of command buffers because the corresponding command buffer call used to be inlined and no command buffers were created. This is now cleaned up and does work as expected with minimal CUDA graph size set to 1 with a flag.

### CODE REVIEW

File: xla/service/gpu/runtime/command_buffer_cmd.h
Comment (ezhulenev) [inline xla/service/gpu/runtime/command_buffer_cmd.h:771]: Instead of taking a reference you can use `std::shared_ptr`, the fact that Thunks stay alive is mostly accidental because of the workaround for cupti+cuda graphs bugs, this will go away once CUDA fix will land

File: xla/service/gpu/gpu_executable.cc
Comment (ezhulenev) [inline xla/service/gpu/gpu_executable.cc:478]: If state is properly shared we don't need any special cases here

File: xla/service/gpu/runtime/cudnn_thunk.h
Comment (ezhulenev) [inline xla/service/gpu/runtime/cudnn_thunk.h:46]: How about returning `std::shared_ptr<LazyDnnGraph>` (or something like this), and whoever gets called first (Thunk or corresponding Command) will initialize it.

Basically command emitter should steal (or share) a state from a thunk

File: xla/service/gpu/runtime/command_buffer_cmd.h
Comment (sergachev) [reply xla/service/gpu/runtime/command_buffer_cmd.h]: Done.

File: xla/service/gpu/gpu_executable.cc
Comment (sergachev) [reply xla/service/gpu/gpu_executable.cc]: Done.

File: xla/service/gpu/runtime/cudnn_thunk.h
Comment (sergachev) [reply xla/service/gpu/runtime/cudnn_thunk.h]: I just added a call to Initialize() to ExecuteOnStream() - does this look reasonable to you? Body of Initialize() is already wrapped in call_once().

**Review by ezhulenev — APPROVED** [review]

**Review by cheshire — APPROVED** [review]

**Comment by cheshire** [comment]
Build layering checks are failing: 

```
[/xla/service/gpu/fusions/cudnn_test.cc:19](:10: error: module //third_party/tensorflow/compiler/xla/service/gpu/fusions:cudnn_test_gpu_a100 does not depend on a module exporting 'third_party/tensorflow/compiler/xla/tests/filecheck.h'
```

**Comment by sergachev** [comment]
Fixed the build file.

**Review by cheshire — APPROVED** [review]

**Comment by sergachev** [comment]
Rebased.

**Review by ezhulenev — APPROVED** [review]

**Comment by sergachev** [comment]
I don't know what's holding this one up; tried to update the build file.

**Pull request closed without merging**
